src/modules/musig/keyagg_impl.h

```diff
@@ -0,0 +1,293 @@
+    ARG_CHECK(secp256k1_memcmp_var(ptr, secp256k1_musig_keyagg_cache_magic, 4) == 0);
+    ptr += 4;
+    secp256k1_ge_from_bytes(&cache_i->pk, ptr);
+    ptr += 64;
+    secp256k1_ge_from_bytes_ext(&cache_i->second_pk, ptr);
+    ptr += 64;
+    memcpy(cache_i->pk_hash, ptr, 32);
+    ptr += 32;
+    cache_i->parity_acc = *ptr & 1;
+    ptr += 1;
+    secp256k1_scalar_set_b32(&cache_i->tweak, ptr, NULL);
+    return 1;
+}
+
+/* Initializes SHA256 with fixed midstate. This midstate was computed by applying
+ * SHA256 to SHA256("KeyAgg list")||SHA256("KeyAgg list"). */
+static void secp256k1_musig_keyagglist_sha256(secp256k1_sha256 *sha) {
+    secp256k1_sha256_initialize(sha);
+
+    sha->s[0] = 0xb399d5e0ul;
+    sha->s[1] = 0xc8fff302ul;
+    sha->s[2] = 0x6badac71ul;
+    sha->s[3] = 0x07c5b7f1ul;
+    sha->s[4] = 0x9701e2eful;
+    sha->s[5] = 0x2a72ecf8ul;
+    sha->s[6] = 0x201a4c7bul;
+    sha->s[7] = 0xab148a38ul;
+    sha->bytes = 64;
+}
+
+/* Computes pk_hash = tagged_hash(pk[0], ..., pk[np-1]) */
+static int secp256k1_musig_compute_pk_hash(const secp256k1_context *ctx, unsigned char *pk_hash, const secp256k1_pubkey * const* pk, size_t np) {
+    secp256k1_sha256 sha;
+    size_t i;
+
+    secp256k1_musig_keyagglist_sha256(&sha);
+    for (i = 0; i < np; i++) {
+        unsigned char ser[33];
+        size_t ser_len = sizeof(ser);
+        if (!secp256k1_ec_pubkey_serialize(ctx, ser, &ser_len, pk[i], SECP256K1_EC_COMPRESSED)) {
+            return 0;
+        }
+        VERIFY_CHECK(ser_len == sizeof(ser));
+        secp256k1_sha256_write(&sha, ser, sizeof(ser));
+    }
+    secp256k1_sha256_finalize(&sha, pk_hash);
+    return 1;
+}
+
+/* Initializes SHA256 with fixed midstate. This midstate was computed by applying
+ * SHA256 to SHA256("KeyAgg coefficient")||SHA256("KeyAgg coefficient"). */
+static void secp256k1_musig_keyaggcoef_sha256(secp256k1_sha256 *sha) {
+    secp256k1_sha256_initialize(sha);
+
+    sha->s[0] = 0x6ef02c5aul;
+    sha->s[1] = 0x06a480deul;
+    sha->s[2] = 0x1f298665ul;
+    sha->s[3] = 0x1d1134f2ul;
+    sha->s[4] = 0x56a0b063ul;
+    sha->s[5] = 0x52da4147ul;
+    sha->s[6] = 0xf280d9d4ul;
+    sha->s[7] = 0x4484be15ul;
+    sha->bytes = 64;
+}
+
+/* Compute KeyAgg coefficient which is constant 1 for the second pubkey and
+ * otherwise tagged_hash(pk_hash, x) where pk_hash is the hash of public keys.
+ * second_pk is the point at infinity in case there is no second_pk. Assumes
+ * that pk is not the point at infinity and that the Y-coordinates of pk and
+ * second_pk are normalized. */
+static void secp256k1_musig_keyaggcoef_internal(secp256k1_scalar *r, const unsigned char *pk_hash, secp256k1_ge *pk, const secp256k1_ge *second_pk) {
+    secp256k1_sha256 sha;
+
+    VERIFY_CHECK(!secp256k1_ge_is_infinity(pk));
+
+    if (!secp256k1_ge_is_infinity(second_pk)
+          && secp256k1_fe_equal(&pk->x, &second_pk->x)
+          && secp256k1_fe_is_odd(&pk->y) == secp256k1_fe_is_odd(&second_pk->y)) {
```

Comment:
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

Not sure how important this check is , but I noticed this branch is not covered while checking `gcov` results. IIUC, the test case to add would be for pk and second_pk have the same x value  but even and odd y values.

Reply:
Good catch! We want everything covered by gcov. I replaced this with `ge_equals_ge_var` which was added after I wrote this code.